internal/strace/strace.go

```diff
@@ -186,8 +190,20 @@ func (r *Result) parseEnterSyscall(syscall, args string) error {
 		if err != nil {
 			return err
 		}
+		writeBuffer := ""
 		match := writePattern.FindStringSubmatch(args)
-		r.recordFileWrite(match[1], bytesWritten)
+		firstQuoteIndex := strings.Index(args, "\"")
+		lastQuoteIndex := strings.LastIndex(args, "\"")
+		if firstQuoteIndex != -1 && lastQuoteIndex != -1 && lastQuoteIndex > firstQuoteIndex {
+			// Save the contents between the first and last quote.
+			writeBuffer = args[firstQuoteIndex+1 : lastQuoteIndex]
+			lastIndexComma := strings.LastIndex(args, ",")
+			if lastQuoteIndex+1 != lastIndexComma {
+				// If the write buffer is truncated there will be an ellipses after the quoted write buffer. Save this as well.
+				writeBuffer += args[lastQuoteIndex+1 : lastIndexComma]
+			}
+		}
+		r.recordFileWrite(match[1], writeBuffer, bytesWritten)
```

Comment:
Since you are effectively copying the contents of all writes during the analysis phase from the strace file into the memory of the analysis binary this is going to grow the worker's memory size a lot (especially for packages writing a lot of data).

The current strace parsing iterates over the file and only reads a line at a time, making it hard for the strace parsing to consume all available memory.

But I suspect we're going to see out-of-memory issues when we deploy this code into production as it currently stands.

Some questions I wonder:

1. Have you done any comparisons about how much memory is consumed during analysis with and without the write buffer capture? (e.g. `/usr/bin/time -v` shows "Maximum resident set size (kbytes)" or calling `https://pkg.go.dev/runtime#ReadMemStats` and logging `MemStats.Alloc`)

2. Is it possible for the writeBuffers to be streamed as they are parsed from the strace file, so memory can be freed as the strace is parsed? (e.g. directly to the bucket, or to tempfiles that are then copied to the bucket later)

Reply:
This is a good point. 

I did some logging for MemStats.Alloc running this on django.
Before we run dynamic analysis for any of the phases: ~8,000,000 bytes
After we run dynamic analysis for all phases (no write buffers saved): ~17,000,000 bytes
After we run dynamic analysis for all phases (saving write buffers): ~50,000,000 bytes

So there is a pretty big difference. 

As for having the writeBuffers be streamed as they are parsed from the strace file and written directly to the bucket, I believe this is possible by using the resultstore directly in strace.go when we process each write syscall. I'm not sure if it's desirable to do so at the strace level of abstraction though since we are planning on changing the log parsing implementation. Currently strace.Parse is called on all of the log files though so we would need to change that if we want to write the writebuffers as they are parsed.

Reply:
Thanks for pointing this out Caleb. Can we perhaps guard this with a flag for now? 

Re streaming: can we easily implement something by adding a simple layer of indirection by using the disk? 

i.e. instead of storing all buffers in memory, just write them to disk and pass around a reference to it. Then the resultstore will simply just need to upload everything from disk instead of memory. 

This wouldn't involve changing the log parsing implementation too much right?

Reply:
Following up here from discussions in chat. We'd also have to write to disk at the strace level otherwise we would need to store the data in memory before all the syscalls are processed. In that case should we just directly write to the bucket?

Reply:
(Don't remember if we resolved this question over our meeting/chat already, but I believe we've resolved this?)

Writing to disk is likely preferable to directly writing to the bucket, as it keeps things a bit more modular. The strace module can just write to os.TempDir(), and store references to that in the result structure.

Reply:
Yes, we discussed in chat. Thanks for following up. Currently taking a lot at the implementation.